File: 2024/2024-06-04-star-formation-simulator/simulate.py
import random
import math
import signal

class Vector:
    def __init__(self, x, y):
        self.x = x
        self.y = y

class Particle:
    def __init__(self, x, y, v, mass):
        self.x = x
        self.y = y
        self.v = v
        self.mass = mass

def rand_num(min_val, max_val):
    return random.uniform(min_val, max_val)

MAX_INITIAL_DISTANCE = 200
INTERVAL_SECONDS = 10e10
NUM_PARTICLES = 10000
G = 6.6743e-11  # real
# G = 6.6743e-6
EARTH_LIKE_MASS_KG = 5.972e24
MOON_LIKE_MASS_KG = 7.348e22
HYDROGEN_MASS_KG = 1.67e-27
COLLISION_DISTANCE_METERS = 2
MINIMUM_FORCE_DISTANCE_METERS = 2

def create_random_particle():
    return Particle(
        x=rand_num(0, 1600),
        y=rand_num(0, 1000),
        v=Vector(0, 0),
        mass=HYDROGEN_MASS_KG * 60e25
    )

def distance(p1, p2):
    dx = p2.x - p1.x
    dy = p2.y - p1.y
    return math.sqrt(dx * dx + dy * dy)

def calculate_gravitational_force(p1, p2):
    dx = p2.x - p1.x
    dy = p2.y - p1.y
    dist = math.sqrt(dx * dx + dy * dy)
    
    if dist < MINIMUM_FORCE_DISTANCE_METERS:
        return Vector(0, 0)
    
    force_magnitude = (G * p1.mass * p2.mass) / (dist * dist)
    force_x = (force_magnitude * dx) / dist
    force_y = (force_magnitude * dy) / dist
    
    return Vector(force_x, force_y)

def apply_force(particle, force, delta_time):
    acceleration_x = force.x / particle.mass
    acceleration_y = force.y / particle.mass
    
    particle.v.x += acceleration_x * delta_time
    particle.v.y += acceleration_y * delta_time

def vector_sum(x, y):
    return x + y

def add_vectors(vectors):
    x_sum = sum([v.x for v in vectors])
    y_sum = sum([v.y for v in vectors])
    return Vector(x_sum, y_sum)

collision_count = 0
particle_count = NUM_PARTICLES
max_mass = 0

def combine_particles(particles):
    global collision_count, particle_count, max_mass
    combined_particles = []
    deleted_indexes = {}

    for i, particle1 in enumerate(particles):
        if i in deleted_indexes:
            continue
        
        for j in range(i + 1, len(particles)):
            if j in deleted_indexes:
                continue
            particle2 = particles[j]
            
            dx = particle1.x - particle2.x
            dy = particle1.y - particle2.y
            dist = math.sqrt(dx * dx + dy * dy)
            
            if dist < COLLISION_DISTANCE_METERS:
                total_mass = particle1.mass + particle2.mass
                collision_count += 1
                particle_count -= 1
                max_mass = max(max_mass, total_mass)
                
                combined_vx = (particle1.mass * particle1.v.x + particle2.mass * particle2.v.x) / total_mass
                combined_vy = (particle1.mass * particle1.v.y + particle2.mass * particle2.v.y) / total_mass
                
                particle1.mass = total_mass
                particle1.v = Vector(combined_vx, combined_vy)
                
                deleted_indexes[j] = True
                break
        
        combined_particles.append(particle1)
    
    return combined_particles

def update_particles(particles):
    for p in particles:
        p.x += p.v.x
        p.y += p.v.y
        
        force_vector = add_vectors([calculate_gravitational_force(p, p2) for p2 in particles])
        apply_force(p, force_vector, INTERVAL_SECONDS)
    
    return combine_particles(particles)

def print_particles(particles):
    for p in particles:
        print(f"{p.x},{p.y},{p.mass}")
    print("---")

particles = [create_random_particle() for _ in range(NUM_PARTICLES)]

interrupted = False

def shutdown(signal, frame):
    global interrupted
    interrupted = True

signal.signal(signal.SIGINT, shutdown)

i = 0
while not interrupted:
    i += 1
    print(i)
    particles = update_particles(particles)
    print_particles(particles)

# Uncomment the line below to shutdown after a specified time
# import threading
# threading.Timer(10, shutdown).start()

File: 2024/2024-06-04-star-formation-simulator/simulate.ts
import type {
  Vector,
  Particle,
  GravitySource,
  ParticleCollection,
} from "./types";
import { ParticleQuadTree } from "./particle-quad-tree";
import { ParticleList } from "./particle-list";
import { ClusterStats } from "./cluster-stats";

const randNum = (min: number, max: number) => {
  return Math.random() * (max - min) + min;
};

const getInt = (key: string): number => {
  const strVal = process.env[key];
  if (strVal == null) {
    throw new Error(`Missing environment variable ${key}`);
  }
  // parseFloat is compatible with scientific notation (e.g. 10e9), while parseInt is not
  const floatVal = parseFloat(strVal);
  if (isNaN(floatVal)) {
    throw new Error(`Expected environment variable ${key} to be an integer`);
  }
  return Math.floor(floatVal);
};

// For reference
// const EARTH_LIKE_MASS_KG = 5.972e24;
// const MOON_LIKE_MASS_KG = 7.348e22;
// const HYDROGEN_MASS_KG = 1.67e-27;

const NURSERY_WIDTH_METERS = getInt("NURSERY_WIDTH_METERS");
const NURSERY_HEIGHT_METERS = getInt("NURSERY_HEIGHT_METERS");
const INTERVAL_SECONDS = getInt("INTERVAL_SECONDS");
const NUM_PARTICLES = getInt("NUM_PARTICLES");
const PARTICLE_MASS_KG = getInt("PARTICLE_MASS_KG");
const ENABLE_GRAVITATIONAL_CLUSTERING = getInt(
  "ENABLE_GRAVITATIONAL_CLUSTERING"
);

const G = 6.6743e-11;
const MIN_COLLISION_DISTANCE_METERS = 1;
const MINIMUM_FORCE_DISTANCE_METERS = 2;

let nextParticleId = 0;

const createRandomParticle = (): Particle => {
  return {
    id: nextParticleId++,
    x: randNum(NURSERY_WIDTH_METERS / -2, NURSERY_WIDTH_METERS / 2),
    y: randNum(NURSERY_HEIGHT_METERS / -2, NURSERY_HEIGHT_METERS / 2),
    v: {
      //   x: randNum(-1, 1),
      //   y: randNum(-1, 1),
      x: 0,
      y: 0,
    },
    mass: PARTICLE_MASS_KG,
  };
};

const calculateGravitationalForce = (
  p1: Particle,
  p2: GravitySource
): Vector => {
  // Calculate the distance vector between p1 and p2
  const dx = p2.x - p1.x;
  const dy = p2.y - p1.y;

  // Calculate the distance magnitude
  const distance = Math.sqrt(dx * dx + dy * dy);

  // Avoid division by zero
  if (distance < MINIMUM_FORCE_DISTANCE_METERS) {
    return { x: 0, y: 0 };
  }

  // Calculate the force magnitude
  const forceMagnitude = (G * p1.mass * p2.mass) / (distance * distance);

  // Calculate the force components
  const forceX = (forceMagnitude * dx) / distance;
  const forceY = (forceMagnitude * dy) / distance;

  return { x: forceX, y: forceY };
};

function applyForce(
  particle: Particle,
  force: Vector,
  deltaTime: number
): void {
  // Calculate the acceleration
  const accelerationX = force.x / particle.mass;
  const accelerationY = force.y / particle.mass;

  // Update the velocity
  particle.v.x += accelerationX * deltaTime;
  particle.v.y += accelerationY * deltaTime;
}

const sum = (x: number, y: number) => x + y;

const addVectors = (vectors: Vector[]): Vector => {
  return {
    x: vectors.map((v) => v.x).reduce(sum, 0),
    y: vectors.map((v) => v.y).reduce(sum, 0),
  };
};

let collisionCount = 0;
let particleCount = NUM_PARTICLES;
let maxMass = 0;

function radiusFromArea(area: number): number {
  return Math.sqrt(area / Math.PI);
}

function combineParticles(pqt: ParticleCollection) {
  const particles = pqt.getAll();
  const deletedParticleIds: Record<number, boolean> = {};

  const distance = particles.reduce((max, p) => {
    const radius = radiusFromArea(p.mass);
    return Math.max(max, radius);
  }, -Infinity);

  for (let i = 0; i < particles.length; i++) {
    let particle1 = particles[i];
    if (deletedParticleIds[particle1.id]) continue;
    const neighbors = pqt.getNeighbors(particle1, distance);
    for (
      let j = particles === neighbors ? i + 1 : 0;
      j < neighbors.length;
      j++
    ) {
      const particle2 = neighbors[j];
      if (deletedParticleIds[particle2.id]) continue;

      const dx = particle1.x - particle2.x;
      const dy = particle1.y - particle2.y;
      const distance = Math.sqrt(dx * dx + dy * dy);
      const collisionDistance = Math.max(
        MIN_COLLISION_DISTANCE_METERS,
        radiusFromArea(particle1.mass) + radiusFromArea(particle2.mass)
      );

      if (distance < collisionDistance) {
        if (particle1.id === particle2.id) {
          throw new Error("these are the same particle");
        }

        pqt.remove(particle1);
        pqt.remove(particle2);

        const cluster = new ClusterStats();
        cluster.addParticle(particle1);
        cluster.addParticle(particle2);

        if (cluster.centerOfMass == null) {
          // just to make type system know this is not null
          throw new Error("Unexpected");
        }

        collisionCount++;
        particleCount--;
        maxMass = Math.max(maxMass, cluster.mass);

        const combinedVx =
          (particle1.mass * particle1.v.x + particle2.mass * particle2.v.x) /
          cluster.mass;
        const combinedVy =
          (particle1.mass * particle1.v.y + particle2.mass * particle2.v.y) /
          cluster.mass;

        const combinedParticle: Particle = {
          id: particle1.mass > particle2.mass ? particle1.id : particle2.id, // preserve id of largest particle. useful for tracking.
          mass: cluster.mass,
          v: { x: combinedVx, y: combinedVy },
          x: cluster.centerOfMass?.x,
          y: cluster.centerOfMass?.y,
        };
        pqt.add(combinedParticle);

        deletedParticleIds[particle1.id] = true;
        deletedParticleIds[particle2.id] = true;
        particle1 = combinedParticle;
      }
    }
  }
}

const updateParticles = (pqt: ParticleCollection) => {
  const particles = pqt.getAll();

  for (const p of particles) {
    pqt.mutateParticle(p, (p: Particle) => {
      p.x += p.v.x * INTERVAL_SECONDS;
      p.y += p.v.y * INTERVAL_SECONDS;
      return p;
    });

    const gravitySources = pqt.getGravitySources(p);
    const forceVector = addVectors(
      gravitySources.map((g) => calculateGravitationalForce(p, g))
    );
    applyForce(p, forceVector, INTERVAL_SECONDS);
  }

  combineParticles(pqt);
};

const printParticles = (particles: Particle[]) => {
  for (let p of particles) {
    console.log(
      [p.id, Math.floor(p.x), Math.floor(p.y), Math.floor(p.mass)].join(",")
    );
  }
  console.log("---");
};

let particleCollection: ParticleCollection;
if (ENABLE_GRAVITATIONAL_CLUSTERING) {
  particleCollection = new ParticleQuadTree();
} else {
  particleCollection = new ParticleList();
}
for (let i = 0; i < NUM_PARTICLES; i++) {
  particleCollection.add(createRandomParticle());
}

let interrupted = false;

let generation = 0;
const start = Date.now();
let lastGenerationRateReport = 0;
while (!interrupted) {
  updateParticles(particleCollection);
  printParticles(particleCollection.getAll());
  generation++;

  if (Date.now() - lastGenerationRateReport > 10000) {
    console.error(
      `generation ${generation}: seconds per generation: ${
        (Date.now() - start) / 1000 / generation
      }`
    );
    lastGenerationRateReport = Date.now();
  }

  // TODO: DON'T LEAVE THIS
  if (generation > 1000) {
    break;
  }
}

const shutdown = () => {
  interrupted = true;
};

process.on("SIGINT", shutdown);
